fix: migrate validator store to event-driven architecture

diegomrsantos opened the pull request:
## Issue Addressed

This fixes the bug where validator state changes during batch processing were not immediately visible to the validator store, as the DB watcher only saw the final state after all transactions completed.

## Proposed Changes

- Add minimal event bus with ValidatorAdded/ValidatorRemoved events
- Update event processor to emit events after successful DB commits
- Refactor validator store to use event bus instead of DB watcher
- Implement incremental event handling for better performance
- Wire up event bus through client and syncer components

## Additional Info

Please provide any additional information. For example, future considerations
or information useful for reviewers.

diegomrsantos committed:
fix: migrate validator store to event-driven architecture

- Add minimal event bus with ValidatorAdded/ValidatorRemoved events
- Update event processor to emit events after successful DB commits
- Refactor validator store to use event bus instead of DB watcher
- Implement incremental event handling for better performance
- Wire up event bus through client and syncer components

This fixes the bug where validator state changes during batch processing
were not immediately visible to the validator store, as the DB watcher
only saw the final state after all transactions completed.

diegomrsantos commented on anchor/eth/src/event_processor.rs:
```diff
@@ -64,6 +67,9 @@ impl EventProcessor {
         let timer = metrics::start_timer(&metrics::EXECUTION_LOG_PROCESSING_TIME);
 
+        // Collect events to emit after successful commit
+        let mut events_to_emit = Vec::new();
+
```
Scale doesn't justify it. SSV events are low-frequency, human-scale operations

diegomrsantos commented on anchor/validator_store/src/lib.rs:
```diff
@@ -154,28 +160,99 @@ impl<T: SlotClock, E: EthSpec> AnchorValidatorStore<T, E> {
+        database_state: watch::Receiver<NetworkState>,
+    ) {
+        // Load initial state from database
+        self.load_validators(&database_state.borrow());
```
@dknopik, the error handling suggestion is incorrect, but how about the `borrow_and_update()`?

dknopik replied:
Do we still need the receiver here? Does a reference to the state suffice now that we have the event bus?

dknopik commented on anchor/validator_store/src/events.rs:
```diff
@@ -0,0 +1,119 @@
+use tracing::info;
+use types::PublicKeyBytes;
+
+/// Events that can occur in the validator lifecycle (minimal set for state sync fix)
+#[derive(Debug, Clone)]
+pub enum ValidatorEvent {
+    /// A validator has been added to the network
+    ValidatorAdded {
+        cluster_id: ClusterId,
+        validator_pubkey: PublicKeyBytes,
+    },
+    /// A validator has been removed from the network  
+    ValidatorRemoved {
+        cluster_id: ClusterId,
+        validator_pubkey: PublicKeyBytes,
+    },
+}
```
Is this the right crate for this module? I would expect this to live with the place that emits the events.

dknopik commented on anchor/validator_store/src/lib.rs:
```diff
@@ -115,6 +118,8 @@ pub struct AnchorValidatorStore<T: SlotClock + 'static, E: EthSpec> {
     builder_boost_factor: Option<u64>,
     prefer_builder_proposals: bool,
     is_synced: watch::Receiver<bool>,
+    /// Event bus for real-time state synchronization during batch processing
+    event_bus: SharedEventBus,
```
This is also for synchronization if we are not batch processing (i.e. in live sync), right?

diegomrsantos replied:
outdated

dknopik commented on anchor/validator_store/src/lib.rs:
```diff
@@ -154,28 +160,98 @@ impl<T: SlotClock, E: EthSpec> AnchorValidatorStore<T, E> {
+        self.load_validators(&database_state.borrow());
+
+        // Then listen for real-time events
+        while let Ok(event) = event_receiver.recv().await {
```
This exits when the receiver returns `RecvError::Lagged`

diegomrsantos commented on anchor/validator_store/src/events.rs:
```diff
@@ -0,0 +1,119 @@
+impl EventBus {
+    /// Create a new event bus
+    pub fn new() -> Self {
+        let (sender, _) = tokio::sync::broadcast::channel(1024);
```
> Lagging
As sent messages must be retained until all Receiver handles receive a clone, broadcast channels are susceptible to the "slow receiver" problem. In this case, all but one receiver are able to receive values at the rate they are sent. Because one receiver is stalled, the channel starts to fill up.
This broadcast channel implementation handles this case by setting a hard upper bound on the number of values the channel may retain at any given time. This upper bound is passed to the channel function as an argument.
If a value is sent when the channel is at capacity, the oldest value currently held by the channel is released. This frees up space for the new value. Any receiver that has not yet seen the released value will return RecvError::Lagged the next time recv is called.
Once RecvError::Lagged is returned, the lagging receiver's position is updated to the oldest value contained by the channel. The next call to recv will return this value.
This behavior enables a receiver to detect when it has lagged so far behind that data has been dropped. The caller may decide how to respond to this: either by aborting its task or by tolerating lost messages and resuming consumption of the channel.

Given the docs above, this channel might not be the right choice.

diegomrsantos committed:
refactor: improve validator store event system

- Replace broadcast channel with MPSC for reliable validator event delivery
- Improve validator store initialization with extracted state
- Add comprehensive event types for full validator lifecycle
- Ensure no event loss with blocking MPSC channel design
- Update event processor to handle new async emit pattern
- Add proper error handling for channel operations
- Maintain backwards compatibility while improving performance

Zacholme7 reviewed (commented):
This looks nice, but I think I am struggling with the idea of adding this whole event bus just for this specific purpose. If our whole client was event drive then this makes sense, but it just feels out of place with the rest of our codebase IMO.

diegomrsantos commented:
> This looks nice, but I think I am struggling with the idea of adding this whole event bus just for this specific purpose. If our whole client was event drive then this makes sense, but it just feels out of place with the rest of our codebase IMO.

If we agree to this architecture, the intention is to utilize it for the entire system.

Zacholme7 commented:
Isn't that a pretty massive change that has to go through testing from the start again? Seems like a decision we should have made awhile ago. Im hesitant to fix something that isn't really broken

diegomrsantos commented:
>  go through testing from the start again?

What do you mean?

> Seems like a decision we should have made awhile ago

We never really discussed many architectural decisions. But I don't see why we can't improve it. We will never get everything right from the beginning. 

> Im hesitant to fix something that isn't really broken

It's actually broken as this is motivated by the current bug in the validator store. Additionally, the current approach in the system is quite complex and error-prone; it might have other bugs.

diegomrsantos commented:
And we don't need to change the entire system at once.

diegomrsantos commented on anchor/validator_store/src/events.rs:
```diff
@@ -0,0 +1,272 @@
+    /// A validator has been added to the network
+    ValidatorAdded {
+        validator_pubkey: PublicKeyBytes,
+        cluster: Box<Cluster>,
```
@dknopik what do you think?

dknopik replied:
Sounds good. Does clippy complain if it is not boxed?

diegomrsantos replied:
Yes, it complains cause the enum is too large. I'm currently using an `Arc` instead.

diegomrsantos commented on anchor/validator_store/src/lib.rs:
```diff
@@ -156,123 +153,80 @@ impl<T: SlotClock, E: EthSpec> AnchorValidatorStore<T, E> {
+            match event {
+                ValidatorEvent::ValidatorAdded {
+                    validator_pubkey,
+                    cluster,
```
This is pattern matching

diegomrsantos committed:
perf: replace Box<Cluster> with Arc<Cluster> to avoid clones

- Replace Box::new(cluster.clone()) with Arc::new(cluster) in event processor
- Update ValidatorEvent::ValidatorAdded to use Arc<Cluster> instead of Box<Cluster>
- Update test cases to use Arc::new instead of Box::new
- Eliminates expensive cluster data cloning during event emission
- Maintains shared ownership semantics with reference counting

diegomrsantos committed:
remove option

diegomrsantos committed:
Merge branch 'unstable' into fix/validator-store-event-driven

diegomrsantos commented:
It was a nice experiment, but this approach is too complex and doesn't add much value to the system. I'm closing it for now.
